### PR TITLE
[diags] support platform specific diags in radio-only ncp

### PR DESCRIPTION
**Pull request by chshu**

### CODE REVIEW

File: src/core/api/link_raw_api.cpp [hunk 1]
```diff
@@ -766,35 +769,87 @@ uint16_t otLinkGetShortAddress(otInstance *aInstance)
...
+    SuccessOrExit(error = Utils::CmdLineParser::ParseCmd(buffer, argCount, argVector, kMaxArgs));
+    VerifyOrExit(argCount >= 1, error = OT_ERROR_INVALID_ARGS);
+
+    if (strcmp(argVector[0], "power") == 0)
+    {
+        char * endptr;
+        int8_t power;
+
+        VerifyOrExit(argCount == 2, error = OT_ERROR_INVALID_ARGS);
+        power = (int8_t)strtol(argVector[1], &endptr, 0);
```
Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: static_cast?

File: src/core/api/link_raw_api.cpp [hunk 2]
```diff
@@ -766,35 +769,87 @@ uint16_t otLinkGetShortAddress(otInstance *aInstance)
...
-    (void)aOutputMaxLen;
+    enum
+    {
+        kMaxArgs          = OPENTHREAD_CONFIG_DIAG_CMD_LINE_ARGS_MAX,
+        kMaxCommandBuffer = OPENTHREAD_CONFIG_DIAG_CMD_LINE_BUFFER_SIZE,
+    };
+
+    otError error = OT_ERROR_NONE;
+    char    buffer[kMaxCommandBuffer];
+    char *  argVector[kMaxArgs];
```
Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: const char*?

File: src/core/api/link_raw_api.cpp [hunk 3]
```diff
@@ -766,35 +769,87 @@ uint16_t otLinkGetShortAddress(otInstance *aInstance)
...
+    {
+    case OT_ERROR_NONE:
+        break;
+
+    case OT_ERROR_NO_BUFS:
+        snprintf(aOutput, aOutputMaxLen, "failed: command string too long\r\n");
+        break;
+
+    default:
+        snprintf(aOutput, aOutputMaxLen, "failed to parse command string\r\n");
```
Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: is it possible to use `otThreadErrorToString()`?

File: src/posix/platform/radio_spinel.cpp [hunk 4]
```diff
@@ -1683,3 +1713,72 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+
+void otPlatDiagProcess(otInstance *aInstance, int argc, char *argv[], char *aOutput, size_t aOutputMaxLen)
+{
+    // delivery the platform specific diags commands to radio only ncp.
+    OT_UNUSED_VARIABLE(aInstance);
+    char        cmd[OPENTHREAD_CONFIG_DIAG_CMD_LINE_BUFFER_SIZE] = {'\0'};
+    const char *space                                            = " ";
+    int         index                                            = 0;
+
+    while (index < argc)
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: use `for`?

File: src/posix/platform/radio_spinel.cpp [hunk 5]
```diff
@@ -1683,3 +1713,72 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+    {
+        strcat(cmd, argv[index]);
+        strcat(cmd, space);
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: space -> `' '`?

File: src/posix/platform/radio_spinel.cpp [hunk 6]
```diff
@@ -1683,3 +1713,72 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+{
+    // delivery the platform specific diags commands to radio only ncp.
+    OT_UNUSED_VARIABLE(aInstance);
+    char        cmd[OPENTHREAD_CONFIG_DIAG_CMD_LINE_BUFFER_SIZE] = {'\0'};
+    const char *space                                            = " ";
+    int         index                                            = 0;
+
+    while (index < argc)
+    {
+        strcat(cmd, argv[index]);
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: it's not safe to use `strcat()`, how about using `snprintf()`?

File: src/posix/platform/radio_spinel.cpp [hunk 7]
```diff
@@ -1683,3 +1713,72 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+
+    sRadioSpinel.PlatDiagProcess(cmd, aOutput, aOutputMaxLen);
+}
+
+void otPlatDiagModeSet(bool aMode)
+{
+    char cmd[OPENTHREAD_CONFIG_DIAG_CMD_LINE_BUFFER_SIZE] = {'\0'};
+
+    sprintf(cmd, aMode ? "start" : "stop");
+    SuccessOrExit(sRadioSpinel.PlatDiagProcess(cmd, NULL, 0));
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: directly use `"start"` or `"stop"`?

File: src/posix/platform/radio_spinel.cpp [hunk 8]
```diff
@@ -1683,3 +1713,72 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+bool otPlatDiagModeGet()
+{
+    return sDiagMode;
+}
+
+void otPlatDiagTxPowerSet(int8_t aTxPower)
+{
+    // other than channel, txpower is not included in radio packet,
+    // so set it to radio only ncp separately.
+    char cmd[OPENTHREAD_CONFIG_DIAG_CMD_LINE_BUFFER_SIZE] = {'\0'};
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: no need to initialize.

File: src/posix/platform/radio_spinel.cpp [hunk 9]
```diff
@@ -1683,3 +1713,72 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
     OT_UNUSED_VARIABLE(aInstance);
 }
 #endif // OPENTHREAD_POSIX_VIRTUAL_TIME
+
+#if OPENTHREAD_ENABLE_DIAG
+static bool sDiagMode = false;
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: move this into `RadioSpinel`?

File: src/core/api/link_raw_api.cpp [hunk 10]
```diff
@@ -766,35 +769,83 @@ uint16_t otLinkGetShortAddress(otInstance *aInstance)
 }
 
 #if OPENTHREAD_ENABLE_DIAG
+otInstance *sDiagInstance;
```
Comment (bukepo) [inline src/core/api/link_raw_api.cpp]: add static?

File: src/posix/platform/radio_spinel.cpp [hunk 11]
```diff
@@ -551,6 +556,15 @@ void RadioSpinel::HandleWaitingResponse(uint32_t          aCommand,
         VerifyOrExit(unpacked > 0, mError = OT_ERROR_PARSE);
         mError = SpinelStatusToOtError(status);
     }
+#if OPENTHREAD_ENABLE_DIAG
+    else if (aKey == SPINEL_PROP_NEST_STREAM_MFG)
+    {
+        VerifyOrExit(mDiagOutput != NULL);
+        spinel_ssize_t unpacked =
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: declare `unpacked` before `VerifyOrExit`?

File: src/posix/platform/radio_spinel.cpp [hunk 12]
```diff
@@ -1692,3 +1723,66 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+{
+    // delivery the platform specific diags commands to radio only ncp.
+    OT_UNUSED_VARIABLE(aInstance);
+    char  cmd[OPENTHREAD_CONFIG_DIAG_CMD_LINE_BUFFER_SIZE] = {'\0'};
+    char *cur                                              = cmd;
+    char *end                                              = cmd + sizeof(cmd);
+
+    for (int index = 0; index < argc; index++)
+    {
+        snprintf(cur, end - cur, "%s ", argv[index]);
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.cpp]: `snprintf()` returns the bytes printed. no need for `strlen()` below.

File: src/posix/platform/radio_spinel.hpp [hunk 13]
```diff
@@ -379,6 +379,38 @@ class RadioSpinel
...
+     */
+    void SetDiagMode(bool aMode) { mDiagMode = aMode; }
+
+    /**
+     * This method indicates whether or not factory diagnostics mode is enabled.
+     *
+     * @returns TRUE if factory diagnostics mode is enabled, FALSE otherwise.
+     *
+     */
+    bool GetDiagMode(void) { return mDiagMode; }
```
Comment (bukepo) [inline src/posix/platform/radio_spinel.hpp]: `const`.

Comment (chshu) [reply src/posix/platform/radio_spinel.hpp]: Updated, thanks.

File: src/posix/platform/radio_spinel.cpp [hunk 14]
```diff
@@ -1699,3 +1732,65 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+    snprintf(cmd, sizeof(cmd), "power %d", aTxPower);
+    SuccessOrExit(sRadioSpinel.PlatDiagProcess(cmd, NULL, 0));
+
+exit:
+    return;
+}
+
+void otPlatDiagChannelSet(uint8_t aChannel)
+{
+    (void)aChannel;
```
Comment (abtink) [inline src/posix/platform/radio_spinel.cpp]: Should we pass it to radio? (like `power` above)

File: src/posix/platform/radio_spinel.cpp [hunk 15]
```diff
@@ -1699,3 +1732,65 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
...
+void otPlatDiagModeSet(bool aMode)
+{
+    SuccessOrExit(sRadioSpinel.PlatDiagProcess(aMode ? "start" : "stop", NULL, 0));
+    sRadioSpinel.SetDiagMode(aMode);
+
+exit:
+    return;
+}
+
+bool otPlatDiagModeGet()
```
Comment (abtink) [inline src/posix/platform/radio_spinel.cpp]: nit: `()` -> `(void)` ?

File: src/core/api/link_raw_api.cpp [hunk 16]
```diff
@@ -766,35 +769,83 @@ uint16_t otLinkGetShortAddress(otInstance *aInstance)
...
+    {
+        char * endptr;
+        int8_t power;
+
+        VerifyOrExit(argCount == 2, error = OT_ERROR_INVALID_ARGS);
+        power = static_cast<int8_t>(strtol(argVector[1], &endptr, 0));
+        VerifyOrExit(*endptr == '\0', error = OT_ERROR_INVALID_ARGS);
+
+        otPlatDiagTxPowerSet(power);
+    }
```
Comment (abtink) [inline src/core/api/link_raw_api.cpp]: How about doing `"channel"` as well (passing it through to `otPlatDiagChannelSet`)?

File: src/posix/platform/radio_spinel.hpp [hunk 17]
```diff
@@ -399,6 +399,38 @@ class RadioSpinel
...
 #endif
 
+#if OPENTHREAD_ENABLE_DIAG
+    /**
+     * This method enables/disables the factory diagnostics mode.
+     *
+     * @param[in]  aMode  TRUE to enable diagnostics mode, FALSE otherwise.
+     *
+     */
+    void SetDiagMode(bool aMode) { mDiagMode = aMode; }
```
Comment (jwhui) [inline src/posix/platform/radio_spinel.hpp]: nit: how about `SetDiagEnabled()` and `IsDiagEnabled()` to better indicate this is a boolean?

File: src/posix/platform/radio_spinel.cpp [hunk 18]
```diff
@@ -1714,3 +1747,69 @@ void otSimRadioSpinelProcess(otInstance *aInstance, const struct Event *aEvent)
     OT_UNUSED_VARIABLE(aInstance);
 }
 #endif // OPENTHREAD_POSIX_VIRTUAL_TIME
+
+#if OPENTHREAD_ENABLE_DIAG
+void otPlatDiagProcess(otInstance *aInstance, int argc, char *argv[], char *aOutput, size_t aOutputMaxLen)
+{
+    // delivery the platform specific diags commands to radio only ncp.
```
Comment (jwhui) [inline src/posix/platform/radio_spinel.cpp]: nit: "delivery" -> "deliver"

**Pull request merged**